### PR TITLE
umd-3.htcondor

### PR DESCRIPTION
**Pull request by guillaume-philippon**
Some additionnal modification to integrate HTCondor into CREAM-CE

### CODE REVIEW

**Comment by guillaume-philippon** [comment]
Please, not merge this pull request yet.

File: features/blparser/config.pan [hunk 1]
```diff
@@ -116,7 +116,9 @@ include { 'components/altlogrotate/config' };
     # Remove potential duplicates
     foreach (lrms;params;BLPARSER_LRMS_PARAMS) {
       if ( index(params['logFile'],log_files) < 0 ) {
-        log_files[length(log_files)] = params['logFile'];
+      	 if(lrms != 'condor' || CE_BATCH_SYS == 'condor' ){
+	         log_files[length(log_files)] = params['logFile'];
+	};
```
Comment (stdweird) [inline features/blparser/config.pan]: fix indentation (2 spaces at least; 4 spaces like everything else if possible)

Comment (jouvin) [inline features/blparser/config.pan]: In pan templates, we historically use 2 spaces... Please keep this for consistency if possible...

File: features/htcondor/params.pan [hunk 2]
```diff
@@ -60,6 +60,16 @@ variable CONDOR_CONFIG = {
     SELF['config.local'] = '';
   };
 
+  # Default configuration is *not* enable Multicore
+  if (!is_defined(SELF['multicore'])){
```
Comment (stdweird) [inline features/htcondor/params.pan]: are these defaults? it's better to make a structire template, and then fill in like
```pan
structure template whatever/htcondordefault;
"a" = "b"; 
```

and then here
```
foreach(k;v;create('whatever/htcondordefault')) {
    if (!is_defined(SELF[k])) { SELF[k]=v;};
}
```

File: features/htcondor/server/groups.pan [hunk 3]
```diff
@@ -69,27 +70,29 @@ variable CONDOR_CONFIG={
   SELF;
 };
 
-'/software/components/filecopy/services/{/usr/libexec/condor_local_submit_attributes.sh}' = {
+include {'components/filecopy/config'};
```
Comment (stdweird) [inline features/htcondor/server/groups.pan]: remove `{}`

**Comment by stdweird** [comment]
i'm not going to pretend i reviewed this, i stopped when i saw filecopy usage :smile:

**Comment by guillaume-philippon** [comment]
It s hard to use metaconfig with condor as you can declare if/else statement and order is important on configuration file.

**Comment by stdweird** [comment]
@guillaume-philippon you'll need TT and possibly complicated schema, but nothing that can't be handled by metaconfig. it's looks similar to `logstash` (or even httpd), which can now be dealt with via metaconfig.
the main issue here is no unittests and no type validation of the values you set via variables; but this is probably out of scope for this PR.

**Comment by jrha** [comment]
@stdweird I'm not entirely sure metaconfig will fit, we've been looking at it at RAL, but the HTCondor configuration has features which allow it to behave as a general purpose language. It might be possible, but very difficult.

**Comment by alahiff** [comment]
We're currently using filecopy for the configuration of our production HTCondor pool. I have a new set of templates for a test pool which is using metaconfig, but without a schema. Using a schema would be rather complicated since:
(1) while the config files are of the form of a list of key-value pairs, the number of keys is huge (judging from what I see doing a condor_config_val at RAL over 1000) since the philosophy is "there's a knob for that"
(2) you're not limited to a pre-defined set of attributes, you can define your own (we have lots of these)
(3) values are not just numbers or strings, they can be expressions, e.g. "SYSTEM_PERIODIC_RELEASE = ((CurrentTime - EnteredCurrentStatus > 30 * 60) && (JobRunCount < 2) && (HoldReasonCode =!= 34))"

**Comment by stdweird** [comment]
@jrha sounds like logstash to me (but possibly worse)
@alahiff 
(1): only support the things you actually use (but if you have no intention of checking anything, then use filecopy by all means)
(2) when people add site customisations, they typically all reinvent the same thing; so it's very likely a good thing to share it via abstract schema.
(3) you can use a string type for that. you can refine the types as you go (expression like that can have errors (`=!=` looks like an ideal candidate :smile:). they make ugly pan schemas, but maintaining this manual is hard too. it depends where you push the work. setting up schema and TT is hard and time consuming once, but then others can use it too)
trying to come up with a schema typically reveals how hard things are to configure.

if htcondor support mutiple config files, or has a load statement to load an external file, you can use that: relative simple things by TT and schema; all weird/complex/manual/hard-to-schemafy on the other files (we do that with moab configs, where part of the config is generated dynamically).

again, i have no clue how complex htcondor is to configure, and if filecopy is the only way to go, fine by me.

**Comment by alahiff** [comment]
HTCondor supports multiple config files (in /etc/condor/config.d). There is also an 'include' statement for including an external file, or the config file can call an arbitrary script which generates additional configuration.

HTCondor configuration can be as simple or as complex as you like :-)

File: features/htcondor/templ/preempt.pan [hunk 4]
```diff
@@ -0,0 +1,12 @@
+structure template features/htcondor/templ/preempt;
+
+'text' = {
+    txt = "";
+    if (CONDOR_CONFIG['wctlimit'] != 0) {
+        txt = txt + "## Configure default wallclocktime ##\n";
+        txt = txt + "PREEMPT = true\n";
+        txt = txt + "MaxJobRetirementTime = " + to_string(CONDOR_CONFIG['wctlimit']) + "\n";
```
Comment (jrha) [inline features/htcondor/templ/preempt.pan]: If you're not going to use metaconfig, then at least use `format()` for what it was designed for...

e.g.
```pan
    template = "## Configure default wallclocktime ##\n";
             + "PREEMPT = true\n"
             + "MaxJobRetirementTime = %f\n";
    txt = format(template, CONDOR_CONFIG['wctlimit']);
```

**Comment by guillaume-philippon** [comment]
I m agree. I will clean the template before merging

**Comment by jrha** [comment]
Cool, thanks :grinning:

**Comment by jrha** [comment]
@guillaume-philippon any update? We're very very close to release now.

**Comment by jrha** [comment]
Ok, good enough for now, will open issue for 15.6 for cleanup.

**Pull request merged**